HIP-1208: Consensus Service Precompiled Contract for Smart Contract Service

walter-hernandez opened the pull request:
**Description**:
<!--
One or two line summary of what this PR does and why it is needed, followed by a list
of changes in imperative, present tense for use in the commit message or changelog. Example:

Add support for ...

* Add config property
* Change column name
* Remove ...
-->

**Related issue(s)**:

Fixes #

**Notes for reviewer**:
<!-- Provide logs, performance numbers or screenshots of the new functionality -->

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

walter-hernandez committed:
Create hip-xxx.md

Signed-off-by: walter-hernandez <52283659+walter-hernandez@users.noreply.github.com>

walter-hernandez committed:
Update and rename hip-xxx.md to hip-1208.md

Uses the PR request number: https://github.com/hiero-ledger/hiero-improvement-proposals/pull/1208

Signed-off-by: walter-hernandez <52283659+walter-hernandez@users.noreply.github.com>

kpachhai commented:
Putting a link to some questions I asked on the discussion post regarding this HIP: https://github.com/hiero-ledger/hiero-improvement-proposals/discussions/1207#discussioncomment-13343498

mgarbs commented on HIP/hip-1208.md:
```diff
@@ -0,0 +1,78 @@
+- The Hedera mainnet provides the trusted consensus layer that ensures all participants within the same topics see the same sequence of events.
+
+## User stories
+- As a developer, I want my smart contract to read data from HCS topics to trigger predefined on-chain actions in a verifiable manner.
```
What would the specification look like for reading topics? How do you envision this working?

walter-hernandez replied:
The HIP has been updated to focus on the writing to topics, leaving the reading from topics as needing another completely different HIP

kpachhai committed:
Updated proposal to include only write operation and added more info with regards to HIP-991 custom fees

Signed-off-by: Kiran Pachhai <kiran.pachhai@hashgraph.com>

kpachhai committed:
Updated the HIP to include additional functions such as createTopic, updateTopic, deleteTopic, getTopicInfo in addition to submitMessage

Signed-off-by: Kiran Pachhai <kiran.pachhai@hashgraph.com>

kpachhai committed:
Updated hedera to hiero in cases where it's relevant

Signed-off-by: Kiran Pachhai <kiran.pachhai@hashgraph.com>

kpachhai committed:
Added usernames to some people in the working group

Signed-off-by: Kiran Pachhai <kiran.pachhai@hashgraph.com>

kpachhai committed:
Added a new person to the working group

Signed-off-by: Kiran Pachhai <kiran.pachhai@hashgraph.com>

walter-hernandez committed:
Update hip-1208.md to reflect authorship contribution and our request

Signed-off-by: walter-hernandez <52283659+walter-hernandez@users.noreply.github.com>

walter-hernandez commented:
HIP updated to reflect authorship of contributors

walter-hernandez committed:
Added requested-by field

Signed-off-by: walter-hernandez <52283659+walter-hernandez@users.noreply.github.com>

ty-swirldslabs commented on HIP/hip-1208.md:
```diff
@@ -1,78 +1,151 @@
 
+Documentation will emphasize the asynchronous nature of the precompile. The recommended workflow is:
+
+1. **Call Precompile**: A developer calls `IHCS(PRECOMPILE_ADDRESS).submitMessage(topic, message, maxHCSTxFeeBudget)` from their contract and verifies the returned `responseCode` is `SUCCESS`.
```
Where would `max_custom_fees` variable go in this function call?

ty-swirldslabs commented on HIP/hip-1208.md:
```diff
@@ -1,78 +1,151 @@
+- The precise gas cost for the precompile call needs to be finalized based on implementation benchmarking.
+- Whether or not to refund the unused `maxHCSTxFeeBudget` back to the user needs to be finalized.
+- Add in Topic (no custom fees) create, update, delete.
+- Add in read Topic info from the state.
```
I would add, How do we implement the `max_custom_fees` variable functionality as defined in this HIP, since it is a departure of current topic functionality

walter-hernandez replied:
We will remove this part about `max_custom_fees` because it is likely to be outside the scope of this HIP

kpachhai replied:
I see it working like this: 
1. Parse Parent Transaction: When a node receives the parent `ContractCallTransaction`, it must parse and store the `max_custom_fees` budget provided by the user.
2. Tag Child Transaction: As the precompile executes and stages a child transaction (like `ConsensusSubmitMessage`), that child must be internally "tagged" with the parent's fee budget context.
3. Enforce Rule Before Charging: When processing the child transaction, the node will compare the topic's required custom fee against the tagged budget. It will apply the new rule for this context so if no budget was tagged, it is treated as zero.
4. Fail on Violation: If the required fee exceeds the budget, the child transaction fails with an appropriate error code (e.g., `CUSTOM_FEE_NOT_AUTHORIZED`), and the user is protected.

I think it makes sense to add this to the proposal. Let me do that to make it clearer.

walter-hernandez replied:
@ty-swirldslabs The possible attack that could exist with the custom fees of topics + the HCS precompile would be that an attacker has a hook on their account that writes to thei custom fees topic. Then, when the attacker receives a transfer to the their account, the hook triggers writing to their custom fees topic, which would mean that the user sending the transfer to the attacker now has to pay the extra fees from the hook writing to the custom fees topic

@kpachhai for 2, what if instead of tagging, the child transaction inherits the parent's fee budget? This could handle a child transaction spawning its own child transactions

Another option to simplify the implementation (removing tagging, bugget tracking, etc.) is to have a hard rule for child transactions to not charge custom fees for topics consumption. This would mean that a contract ID would have to be set on `FeeExemptKeyList` for a topic. Then, a developer can implement whatever fee model makes sense for their use case at the smart contract level, not topic level. The fees would be visible in the contract code and approved by the user when interacting with the smart contract. @ty-swirldslabs @kpachhai what do you think?

ty-swirldslabs reviewed (commented):
This is a great update! I've added some questions and comments for some clarity

walter-hernandez commented on HIP/hip-1208.md:
```diff
@@ -1,78 +1,151 @@
+- **Asynchronous Confirmation**: A `SUCCESS` response from the precompile only confirms the message was staged. Finality must be confirmed by finding the child transaction record off-chain.
+- **Interaction with Custom Topic Fees (HIP-991)**: The introduction of custom fees on HCS topics has significant security implications for this precompile, especially when used with automated systems like EVM Hooks.
+  - **Risk of Unexpected Charges**: A malicious actor could install a hook on their account that calls this precompile to write to a topic with an exorbitant custom fee. An unsuspecting user interacting with that account (e.g., via a simple `CryptoTransfer`) could be exposed to this fee without their consent.
+  - **Required Network Safeguard**: To prevent this, the network protocol must enforce that a child transaction cannot assess a custom fee against a payer unless that payer has explicitly budgeted for it in the parent transaction's `max_custom_fees` field. In the absence of this explicit consent in the parent transaction, the child transaction must fail, protecting the user from non-consensual ch ...
```
We can remove this part about `max_custom_fees` because it is likely to be outside the scope of this HIP. With the custom fixed fees of topics functionality (https://hips.hedera.com/hip/hip-991), it should be enough to limit the fees that someone interacting with a smart contract reading a topic would be subject to

kpachhai replied:
The proposal that an unset `max_custom_fees` budget defaults to zero is a deliberate design choice for user safety, and it is scoped exclusively to child transactions generated by this HSCS precompile.

We are NOT proposing a foundational change to how custom fees work on all native transactions.

The reason for this "safe anti-pattern" is the unique risk profile of the EVM especially with hooks. For example, a user making a simple token transfer has no way of knowing that a hook might trigger an action with a high custom fee. This localized safety measure will protect users in the smart contract environment from unexpected and non-consensual charges. This sort of risk doesn't exist for direct, native transactions.

kpachhai commented on HIP/hip-1208.md:
```diff
@@ -1,78 +1,151 @@
+
+To maintain security and prevent unauthorized interaction with HCS topics:
+
+- **Permission Control**: If a topic has a `submitKey`, a contract can only write to it if the contract's own ID(e.g., 0.0.1234) is part of that `submitKey`. Authorization is tied to the contract itself, not the user calling the contract.
```
The act of the contract calling the precompile serves as its "signature" or authorization. Just to clarify, we don't need to expand or change any functionality. This is already supported by Hedera's native Key structure. A Hedera Key can be a standard public key, but it can also be set directly to a contractID (e.g., 0.0.12345). This also works perfectly with threshold keys, allowing for flexible permissions like a "1-of-N" key that includes the contract ID and other admin keys.

jsync-swirlds replied:
It should be [Contract ID](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hapi/hedera-protobuf-java-api/src/main/proto/services/basic_types.proto#L271) which is the [Key](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hapi/hedera-protobuf-java-api/src/main/proto/services/basic_types.proto#L909) for a contract.
Contracts do not have private keys available and cannot actually sign transactions cryptographically.

walter-hernandez committed:
Clarify use of Contract ID in `submitKey` list

Signed-off-by: walter-hernandez <52283659+walter-hernandez@users.noreply.github.com>

walter-hernandez commented on HIP/hip-1208.md:
```diff
@@ -1,78 +1,151 @@
+- **Message Size and Content**: The precompile will enforce HCS message size limits(e.g. 1024 bytes).
+- **Topic Integrity**: Contracts will not be able to delete or modify existing HCS topics. The creation of HCS topics will still be done through existing HCS APIs via Hiero SDK.
+- **Fee Management**: The `maxHCSTxFeeBudget` must be budgeted carefully. If set too low for the message size, the child transaction will fail post-consensus, even if the parent transaction succeeded.
+- **Asynchronous Confirmation**: A `SUCCESS` response from the precompile only confirms the message was staged. Finality must be confirmed by finding the child transaction record off-chain.
```
Closing this as resolved

kpachhai committed:
Refine specification based on the feedback and comments, simplifying the security model for custom fees. This commit also clarifies key management and on-chain confirmation terminology

Signed-off-by: Kiran Pachhai <kiran.pachhai@hashgraph.com>

ty-swirldslabs reviewed (commented):
These updates LGTM

jsync-swirlds commented on HIP/hip-1208.md:
```diff
@@ -0,0 +1,189 @@
+- **createTopic, updateTopic**: These functions explicitly use the `TopicCreatePayload` struct, which prevents developers from attempting to set custom fee schedules. This enforces the scope of the HIP at the contract interface level. It stages the corresponding child transactions(ConsensusCreateTopic, etc). The `adminKey` and `submitKey` can be set to the contract’s own address(`address(this)`)  ...
+- **deleteTopic**: This function is authorized by the topic’s `adminKey`. It can delete a topic regardless of whether that topic has an existing custom fee schedule.
+- **getTopicInfo**: This is a synchronous read-only call that returns the topic’s metadata. It does not generate a child transaction. It returns the comprehensive `TopicInfo` struct, providing read-only access to all topic properties, including the full custom fee schedule. This allows smart contracts to become “fee-aware”.
+- **submitMessage**: This function can submit messages to any topic, including those with custom fees. It also supports automatic message chunking. If the message payload is larger than the ~6KiB single transaction limit, the precompile will automatically calculate the number of chunks required, validate the `maxHCSTxFeeBudget` against the total cost for all chunks, and stage N child transactions ...
```
It might be better to specify the network property `consensus.message.maxBytesAllowed` as it's configurable and might be different for different Hiero networks.
The default is _currently_ 1024 bytes.

walter-hernandez replied:
@ty-swirldslabs @jsync-swirlds Updated the PR with the network property `consensus.message.maxBytesAllowed`

walter-hernandez committed:
Merge branch 'hiero-ledger:main' into main

walter-hernandez committed:
Fix submitMessage function description for limit

Updated the description of the submitMessage function to reflect the correct transaction limit

Signed-off-by: walter-hernandez <52283659+walter-hernandez@users.noreply.github.com>

mgarbs committed:
chore: reconcile HIP-1208 status

Signed-off-by: Michael Garber <michael.garber@hashgraph.com>
